src/v/cloud_roles/refresh_credentials.cc

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_roles/refresh_credentials.h"
+
+#include "cloud_roles/logger.h"
+#include "config/node_config.h"
+#include "model/metadata.h"
+#include "s3/client.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/loop.hh>
+
+/// These environment variables can be used to override the default hostname and
+/// port for fetching temporary credentials for testing.
+struct override_api_endpoint_env_vars {
+    static constexpr std::string_view host = "RP_SI_CREDS_API_HOST";
+    static constexpr std::string_view port = "RP_SI_CREDS_API_PORT";
+};
+
+/// Multiplier to derive sleep duration from expiry time. Leaves 0.1 * expiry
+/// seconds as buffer to make API calls. For default expiry of 1 hour, this
+/// results in a fetch after 54 minutes.
+static constexpr float sleep_from_expiry_multiplier = 0.9;
+
+cloud_roles::refresh_credentials::refresh_credentials(
+  std::unique_ptr<impl> impl,
+  ss::gate& gate,
+  ss::abort_source& as,
+  credentials_update_cb_t creds_update,
+  s3::aws_region_name region)
+  : _impl(std::move(impl))
+  , _gate(gate)
+  , _as(as)
+  , _credentials_update(std::move(creds_update))
+  , _region{std::move(region)} {}
+
+void cloud_roles::refresh_credentials::start() {
+    ssx::background = ssx::spawn_with_gate_then(
+      _gate, [this]() mutable { return do_start(); });
+}
+
+ss::future<> cloud_roles::refresh_credentials::do_start() {
+    return ss::do_until(
+      [this] { return _gate.is_closed() || _as.abort_requested(); },
+      [this] { return fetch_and_update_credentials(); });
+}
+
+static std::optional<ss::sstring>
+load_and_validate_env_var(std::string_view env_var) {
+    char* override_maybe = std::getenv(env_var.data());
+    if (override_maybe) {
+        ss::sstring override{override_maybe};
+        if (!override.empty()) {
+            return override;
+        }
+
+        vlog(
+          cloud_roles::clrl_log.warn,
+          "override environment variable {} is set but empty, ignoring",
+          env_var);
+    }
+    return std::nullopt;
+}
+
+cloud_roles::refresh_credentials::impl::impl(
+  ss::sstring api_host,
+  uint16_t api_port,
+  s3::aws_region_name region,
+  ss::abort_source& as,
+  retry_params retry_params)
+  : _api_host{std::move(api_host)}
+  , _api_port{api_port}
+  , _region{std::move(region)}
+  , _as{as}
+  , _retry_params{retry_params} {
+    if (config::node().developer_mode()) {
```

Comment:
some context from [this discussion](https://docs.google.com/document/d/1c9KLcyIG8Zjdye7uuUwIJpnwtUgHYLpdJE0Q0LYBb1Y/edit?disco=AAAAZxLelKw) on the rfc (it does not directly mention developer mode but we discussed in calls around this) - the idea was to be able to override the credentials api url and port with env variables, and protect this override with developer mode to avoid user accidentally override the endpoint. 

however if the flag is being removed it makes sense to not use the dev mode here in this context.

Reply:
fixed

Reply:
I see I didn't realize it was agreed to before hand. It's very non-standard, so adding a comment explaining would be useful if you end up leaving it in place.

Reply:
I have removed it now.